tests/src/unit/OauthTest.php

```diff
@@ -0,0 +1,171 @@
+
+    $container = new ContainerBuilder();
+    $container->set('url_generator', $this->urlGenerator);
+    $container->set('string_translation', $this->getStringTranslationStub());
+    \Drupal::setContainer($container);
+
+    $this->oAuth = new Oauth($this->getConfigFactoryStub(), $this->csrfTokenGenerator, $this->urlGenerator, $this->guzzle_client);
+
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getConfigFactoryStub(array $configs = []) {
+    return parent::getConfigFactoryStub([
+      'media_webdam.settings' => [
+        'username' => 'WDusername',
+        'password' => 'WDpassword',
+        'client_id' => 'WDclient-id',
+        'secret' => 'WDsecret',
+      ],
+    ]);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function testGetAuthLink() {
+    $authUrl = $this->oAuth->getAuthLink();
+
+    $this->assertContains('some/url/test', $authUrl);
+    $this->assertContains('testToken112233', $authUrl);
+    $this->assertContains('WDclient-id', $authUrl);
+    $this->assertContains('/oauth2/authorize', $authUrl);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function testAuthRequestStateIsValid() {
+    $token = 'testToken112233';
+    $this->csrfTokenGenerator->expects($this->any())
+      ->method('validate')
+      ->with($token)
+      ->willReturn(TRUE);
+
+    $this->oAuth->authRequestStateIsValid($token);
+    $this->assertTrue($this->csrfTokenGenerator->validate($token));
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function testGetAccessToken($auth_code = '') {
+    $mock = new MockHandler([
+      new Response(200, [], '{"access_token":"ACCESS_TOKEN", "token_type":"bearer", "expires_in":3600, "refresh_token": "refresh_token"}'),
+      new Response(200, [], '{"response_type":"code", "state":"z09GqyCtJFZa-BT2Lz1K_E5ngfCvObZqpGHJyjtlSzc"}'),
+    ]);
+    $handler = HandlerStack::create($mock);
+    $this->httpClient = new GClient(['handler' => $handler]);
+
+    $auth_code = 'somedummycode123';
+
+    $this->oAuth = new Oauth($this->getConfigFactoryStub(), $this->csrfTokenGenerator, $this->urlGenerator, $this->httpClient);
+    $getAccessToken = $this->oAuth->getAccessToken($auth_code);
+
+    $this->assertArrayHasKey('expire_time', $getAccessToken);
+    $this->assertArrayHasKey('access_token', $getAccessToken);
+    $this->assertNotEmpty($getAccessToken['access_token']);
+
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function testSetAuthFinishRedirect($authFinishRedirect = '') {
+    $mock = new MockHandler([
+      new Response(200, [], '{"access_token":"ACCESS_TOKEN", "token_type":"bearer", "expires_in":3600, "refresh_token": "refresh_token"}'),
+      new Response(200, [], '{"auth_finish_redirect":"https://thelongurl.with.useful?parameters"}'),
+    ]);
+    $handler = HandlerStack::create($mock);
+    $this->httpClient = new GClient(['handler' => $handler]);
+
+    $this->oAuth = new Oauth($this->getConfigFactoryStub(), $this->csrfTokenGenerator, $this->urlGenerator, $this->httpClient);
```

Comment:
I think you might need to call one of the methods here still & make assertions about the output.